### PR TITLE
Bug 1131773 - Add an addon-details panel to the Settings app r=eragonj

### PR DESCRIPTION
**Pull request by crh0716**
- Include system apps, homescreens, etc. in the list of target apps
- Figure out whether we need to reboot or restart apps after disabling an addon
- Show a toast to remind user to restart or reboot when disabling an add-on

### CODE REVIEW

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f11165ee8597e4f5c1b79cc4954ecba24a1c451e

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e42fc9ffafcc00f8533c08da98226c22ab026ad4

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7a346ae1b7d9869afdf4b917d93edb17167ec96d

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=403e8802f68e0139c80c5ff39dc4a706ebebe798

File: apps/settings/js/modules/addon_manager.js [hunk 1]
```diff
@@ -0,0 +1,403 @@
+/* global MozActivity */
```
Comment (EragonJ) [inline apps/settings/js/modules/addon_manager.js]: Can't we use window.MozActivity ?

File: apps/settings/js/modules/addon_manager.js [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+     *
+     * @access public
+     * @memberOf AddonManager.prototype
+     * @param {DOMApplication} addon
+     */
+    enableAddon: function(addon) {
+      if (this._isAddon(addon)) {
+        mozApps.mgmt.setEnabled(addon, true);
+      } else {
+        return Promise.reject('not an addon');
```
Comment (EragonJ) [inline apps/settings/js/modules/addon_manager.js]: Why we return Promise here ?? I can't see any promise chain when using it.

File: apps/settings/js/modules/addon_manager.js [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+
+    /**
+     * Enable an addon
+     *
+     * @access public
+     * @memberOf AddonManager.prototype
+     * @param {DOMApplication} addon
+     */
+    enableAddon: function(addon) {
+      if (this._isAddon(addon)) {
```
Comment (EragonJ) [inline apps/settings/js/modules/addon_manager.js]: Maybe we can use `isEnabled` here to do a early return ?

File: apps/settings/js/modules/addon_manager.js [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+     * Disable an addon. Returns a string indicating whether reboot or restart
+     * required for disabling the addon completely.
+     *
+     * @access public
+     * @memberOf AddonManager.prototype
+     * @param {DOMApplication} addon
+     * @returns {Promise String}
+     */
+    disableAddon: function(addon) {
+      if (this._isAddon(addon)) {
```
Comment (EragonJ) [inline apps/settings/js/modules/addon_manager.js]: If we are disabling an addon that is currently enabling, is there any problem or error that we should handle here ?

File: apps/settings/js/modules/addon_manager.js [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+      } else {
+        return Promise.reject('not an addon');
+      }
+
+      return this._getCustomizedApps(addon).then((customizedApps) => {
+        var needsReboot = customizedApps.some((customizedApp) => {
+          var manifest = this._getManifest(customizedApp.app);
+          var role = manifest.role;
+          // If the addon affects a system app
+          if (role === 'system' || role === 'homescreen') {
```
Comment (EragonJ) [inline apps/settings/js/modules/addon_manager.js]: I think we should use create a function with a pre-defined whitelist to make it more easy to maintain !

File: apps/settings/js/modules/addon_manager.js [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+
+    /**
+     * Delete an addon.
+     *
+     * @access public
+     * @memberOf AddonManager.prototype
+     * @param {Object} addon
+     * @returns {Promise}
+     */
+    deleteAddon: function(addon) {
```
Comment (EragonJ) [inline apps/settings/js/modules/addon_manager.js]: If we are disabling an addon that is currently enabling, is there any problem or error that we should handle here ?

File: apps/settings/js/panels/addon_details/addon_details.js [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    this.panel = panel;
+    this.body = panel.querySelector('.addon-details-body');
+    this.header = panel.querySelector('.addon-details-header');
+    this.icon = panel.querySelector('.addon-details-icon');
+    this.description = panel.querySelector('.addon-description-text');
+    this.developer = panel.querySelector('.addon-developer');
+    this.targetsList = panel.querySelector('.addon-targets');
+    this.noTargetsMsg = panel.querySelector('.addon-no-targets');
+    this.toggle = panel.querySelector('.addon-enabled');
+    this.deleteButton = panel.querySelector('.addon-delete');
```
Comment (EragonJ) [inline apps/settings/js/panels/addon_details/addon_details.js]: I think we should keep them inside `_elements` to the code consistent ?

Comment (EragonJ) [inline apps/settings/js/panels/addon_details/addon_details.js]: Take `this.developer` for example, we can't easily recognize its use at the first glimpse.

Comment (EragonJ) [inline apps/settings/js/panels/addon_details/addon_details.js]: for me this part looks more like UI manipulation, should we move this into `addon_details/panel.js` ?

File: apps/settings/js/panels/addon_details/addon_details.js [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+   * @param {DOMApplication} app
+   */
+  AddonDetails.prototype.render = function render(app) {
+    var self = this;
+    var l10n = navigator.mozL10n;
+    this.app = app;
+    var manifest = new ManifestHelper(app.manifest || app.updateManifest);
+
+    // Scroll back to the top
+    this.body.scrollTop = 0;
```
Comment (EragonJ) [inline apps/settings/js/panels/addon_details/addon_details.js]: This looks more like cleanup works for me, maybe we can directly do this in `onHide()` ?

File: apps/settings/js/panels/addon_details/addon_details.js [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+      this.developer.hidden = false;
+      l10n.setAttributes(this.developer, 'addon-developer', {
+        developerName: developerName
+      });
+    } else {
+      this.developer.hidden = true;
+    }
+
+    // Get the list of targeted apps
+    this.targetsList.textContent = '';  // Clear any old content here
```
Comment (EragonJ) [inline apps/settings/js/panels/addon_details/addon_details.js]: same here

File: apps/settings/js/panels/addon_details/addon_details.js [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    this.toggle.checked = AddonManager.isEnabled(app); // set initial state
+    this.toggle.onchange = function() {                // handle user changes
+      if (self.toggle.checked) {
+        AddonManager.enableAddon(app);
+      } else {
+        AddonManager.disableAddon(app).then(function(status) {
+          // If an addon injected a script then we have to restart any
+          // affected apps. And if those apps are system apps, then we
+          // need to reboot the device.
+          if (status === 'reboot') {
```
Comment (EragonJ) [inline apps/settings/js/panels/addon_details/addon_details.js]: For these two cases, should we do anything here to make sure we do reboot / restart related apps ? Or system / Gecko would handle this for us when disabling them ?

File: apps/settings/js/panels/addon_details/addon_details.js [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+  /**
+   * Render details of an app to the UI elements.
+   *
+   * @access public
+   * @memberOf AddonDetails.prototype
+   * @param {DOMApplication} app
+   */
+  AddonDetails.prototype.render = function render(app) {
+    var self = this;
```
Comment (EragonJ) [inline apps/settings/js/panels/addon_details/addon_details.js]: There is no strong reason here that we should use `self`, can we just use arrow function in ES6 to achieve the same thing ?

File: apps/settings/js/panels/addon_details/addon_details.js [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+      this.developer.hidden = true;
+    }
+
+    // Get the list of targeted apps
+    this.targetsList.textContent = '';  // Clear any old content here
+    AddonManager.getAddonTargets(app).then(function(targets) {
+      var names = [];
+      targets.forEach(function(target) {
+        var manifest = new ManifestHelper(target.manifest ||
+                                          target.updateManifest);
```
Comment (EragonJ) [inline apps/settings/js/panels/addon_details/addon_details.js]: wrong spaces ?

File: apps/settings/js/panels/addon_details/addon_details.js (hunk 10 above)
Comment (EragonJ) [inline apps/settings/js/panels/addon_details/addon_details.js]: If we are going to reboot, is there any chance that we would jump to this block and show the toast ?

File: apps/settings/js/panels/addon_details/panel.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+    return SettingsPanel({
+      onInit: function(panel) {
+        details = AddonDetails(panel);
+      },
+
+      onBeforeShow: function(panel, options) {
+        details.render(options.addon);
+      }
+    });
+  };
```
Comment (EragonJ) [inline apps/settings/js/panels/addon_details/panel.js]: XDD I still remember the time when I was writing the first settings module like this !

File: apps/settings/js/panels/addon_details/addon_details.js (hunk 10 above)
Comment (EragonJ) [inline apps/settings/js/panels/addon_details/addon_details.js]: ok, i just flashed the patch and noticed that users have to reboot by themselves.

File: apps/settings/js/modules/addon_manager.js [hunk 14]
```diff
@@ -0,0 +1,403 @@
...
+      if (!this._isAddon(addon)) {
+        return Promise.reject('not an addon');
+      }
+
+      if (!addon.removable) {
+        return Promise.reject('addon is not deletable');
+      }
+
+      return new Promise(function(resolve, reject) {
+        var request = mozApps.mgmt.uninstall(addon);
```
Comment (EragonJ) [inline apps/settings/js/modules/addon_manager.js]: It seems that this line would create a special alert screen. interesting

File: apps/settings/js/modules/addon_manager.js (hunk 1 above)
Comment (crh0716) [inline apps/settings/js/modules/addon_manager.js]: good idea

File: apps/settings/js/modules/addon_manager.js (hunk 2 above)
Comment (crh0716) [inline apps/settings/js/modules/addon_manager.js]: Just align to `disableAddon`.

File: apps/settings/js/modules/addon_manager.js (hunk 3 above)
Comment (crh0716) [inline apps/settings/js/modules/addon_manager.js]: hmm, I think so.

Comment (crh0716) [inline apps/settings/js/modules/addon_manager.js]: Add additional check on `addon.enabled`.

File: apps/settings/js/modules/addon_manager.js (hunk 6 above)
Comment (crh0716) [inline apps/settings/js/modules/addon_manager.js]: As far as I know we don't need to do anything.

File: apps/settings/js/modules/addon_manager.js (hunk 5 above)
Comment (crh0716) [inline apps/settings/js/modules/addon_manager.js]: Let's move the logic to a separate function.

File: apps/settings/js/panels/addon_details/addon_details.js (hunk 7 above)
Comment (crh0716) [inline apps/settings/js/panels/addon_details/addon_details.js]: UI manipulation code can also be a separate module. I'll move operations to `panel.js`.

File: apps/settings/js/panels/addon_details/addon_details.js (hunk 8 above)
Comment (crh0716) [inline apps/settings/js/panels/addon_details/addon_details.js]: :100:

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7292580c0533dd66ff366d562812a15bac6cba12

File: apps/settings/js/modules/addon_manager.js (hunk 6 above)
Comment (EragonJ) [inline apps/settings/js/modules/addon_manager.js]: okay !

File: apps/settings/js/panels/addon_details/addon_details.js (hunk 11 above)
Comment (EragonJ) [inline apps/settings/js/panels/addon_details/addon_details.js]: :100:

File: apps/settings/style/apps.css [hunk 15]
```diff
@@ -38,7 +38,40 @@
...
+ */
+html[dir="rtl"] #addons ul li img,
+html[dir="rtl"] #addon-details ul li img,
+html[dir="rtl"] #appPermissions ul li img,
+html[dir="rtl"] #homescreens ul li img {
+  left: unset;
+  right: 1.6rem;
+}
+
+html[dir="rtl"] .addon-paragraph img {
```
Comment (EragonJ) [inline apps/settings/style/apps.css]: I remembered that there is one mailing thread talking about `[dir="xx"]` and `:dir()` and the latter one is recommended. Can we use `html:dir(rtl)` instead here ? 

Reference : https://developer.mozilla.org/en-US/docs/Web/CSS/:dir

Comment (crh0716) [inline apps/settings/style/apps.css]: sure~

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=290b44027a9f19e31b5cbeb4141c45354e4157e7

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9858fd35ebccd091b49e9f854f9d584dad2f6dfe

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=cfe316caed600d6d83f803beb3d42768a40c60db

**Comment by autolander** [comment]
http://docs.taskcluster.net/tools/task-graph-inspector/#r-akC3F0S6OM9GXGcRAxuw

The pull request failed to pass integration tests. It could not be landed, please try again.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=3463ef39c9dfb17dd1f2587495b30e508bf7cc63

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=495451c59fe513f49b022a48b1c820ea9f645d76

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=38715c5db6c9733219e87047bc19a6409680e0c0

**Comment by autolander** [comment]
http://docs.taskcluster.net/tools/task-graph-inspector/#P9S2UIYRTKeJBuO5VZ35Vg

The pull request failed to pass integration tests. It could not be landed, please try again.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=69af578d8299d8bee3e5aa13b7836d3ad1470c1f

**Comment by autolander** [comment]
http://docs.taskcluster.net/tools/task-graph-inspector/#YnV52bIVQfSZ341JLgy_xg

The pull request failed to pass integration tests. It could not be landed, please try again.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=1379d200622a216ef554b3e06989a129ee5cf9c9

**Pull request merged**